### ISSUE TITLE
Large Data Performance FieldArrayType

### ISSUE THREAD

**Issue by thorgod** [ngx-formly/ngx-formly#1489]
**I'm submitting a ...**  (check one with "x")
```
[ ] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[X] support request
```

**Current behavior**
Slow loading lots of fields/data with FieldArrayType
~72 Field Columns * ~36 Rows takes around 20 seconds. 

**Expected behavior**
Faster loading time on large data.

**Minimal reproduction of the problem with instructions**
(This is small example I whipped up might not be the right one, I will work on getting and example exactly like the problem I am seeing in production)
https://stackblitz.com/edit/angular-pbbsxv?file=src%2Fapp%2Fapp.component.html

Here is an example of the loading time in production of a large form. I am working on creating an stackblitz example that represents this. 
![image](https://user-images.githubusercontent.com/13910170/55038615-ade04180-4fde-11e9-8664-343f46a4de3f.png)

I will be working on also solving this issue... more information to come.

**Comment by aitboudad** [comment]
There two part to explore which would improve the performance:

1. `OnPush` changeDetection: https://github.com/ngx-formly/ngx-formly/issues/1272#issuecomment-437673377
2. use virtual list

**Comment by thorgod** [comment]
Thanks @aitboudad , I will look into those options.

I am working on making an example with large amount of fake data that emulates the problem I am seeing in production, should be done by tomorrow.

**Comment by aitboudad** [comment]
if time allows I'll try to re-check on my side tomorrow

**Comment by thorgod** [comment]
For my example I am working on I use formly-group instead of formly-field for the repeat section custom type. (AKA Table) I do not think it matters but wanted to make sure. 

And the OnPush did not seem to speed things up, on my code base, just tested. 

(But let me get that stackblitz complete example so we are all looking at the same thing I am seeing)

**Comment by thorgod** [comment]
@aitboudad 

I was able to strip everything out and make a basic example. You will see the load time performance on the stackblitz example.


https://stackblitz.com/edit/angular-ngx-formly-large-load?file=src%2Fapp%2Fapp.component.ts

I will start trying to figure out the what the cause is now, but my application has validation fields and other things besides the basics.

**Comment by aitboudad** [comment]
could try to comment [FormlyForm::ngDoCheck](https://github.com/ngx-formly/ngx-formly/blob/master/src/core/src/lib/components/formly.form.ts#L69) and see if you'll notice some improvement

**Comment by aitboudad** [comment]
got it, replaced `form.setControl` in ([field-form.ts#L91](https://github.com/ngx-formly/ngx-formly/blob/master/src/core/src/lib/extensions/field-form/field-form.ts#L91), [field-form.ts#L95](https://github.com/ngx-formly/ngx-formly/blob/master/src/core/src/lib/extensions/field-form/field-form.ts#L95)) calls which seems too expansive and the `buildForm` time decreased significantly from `700ms` to `152ms`.
```patch
    if (form instanceof FormArray) {
      if (form.at(<number> path) !== control) {
--        form.setControl(<number> path, control);
++        form.controls[path] = control;
      }
    } else {
      if (form.get(<string> path) !== control) {
--        form.setControl(<string> path, control);
++        form.controls[path] = control;
      }
    }
```

**Comment by thorgod** [comment]
@aitboudad 
No real difference disabling ngDoCheck, (Also I have OnPush changeDetection in this stack blitz) 

After:
https://stackblitz.com/edit/angular-ngx-formly-large-load-v2
I added ngx-formly to stackblitz and disabled ngDoCheck. 20 Seconds to load form (After data is loaded).

![image](https://user-images.githubusercontent.com/13910170/55259185-3995e000-5222-11e9-8491-e0915c80d7ba.png)


Before: (Seems to be faster, don't know why)
![image](https://user-images.githubusercontent.com/13910170/55259485-0273fe80-5223-11e9-95c3-a9a0718247bb.png)

**Comment by thorgod** [comment]
Wow so much faster. First try. Need to look deeper but first try seems to be a lot faster.

**Comment by thorgod** [comment]
Here is the updated example, with your changes. (FYI)
https://angular-ngx-formly-large-load-v3.stackblitz.io/

I am going to attempt to bring those changes into my application and test it with real data.

**Comment by aitboudad** [comment]
> Note: how did you figure it out? Any tooling you used?

just chrome devtools, the performance graph shows me that `setControl` is the most called ones during `_buildForm` and took an average of `2ms` for fieldGroup type.

> I am going to attempt to bring those changes into my application and test it with real data.

just for checking, but the fix still need some extra work in order to be ready for production use.

**Comment by thorgod** [comment]
I understand, just testing it out on dev. (We have local, dev, test, stage, prod, also everything requires regression and validation before we release lol so it wont be in prod for at least a few months)

**Comment by thorgod** [comment]
OK so no over difference with the changes, in the application; however, it looks like the first part is faster. 

I will look deeper into it, but overall yes it was faster in my example, but when testing on the real app it was still slow. I will add more the stackblitz example maybe something else is the cause.

Before:
![image](https://user-images.githubusercontent.com/13910170/55262663-bda09580-522b-11e9-9715-c67ed06202c3.png)

After:
![image](https://user-images.githubusercontent.com/13910170/55262782-238d1d00-522c-11e9-9942-e3bf73f5b057.png)

**Comment by aitboudad** [comment]
@thorgod check:

- if `formly-form` is used inside a custom field type, if so, replace it with `formly-field` component.
- if you call `add` method during build process of field array or comment the following line https://github.com/ngx-formly/ngx-formly/blob/master/src/core/src/lib/templates/field-array.type.ts#L46

**Comment by thorgod** [comment]
@aitboudad 
Thanks
No we dont have any to formly-form in a custom field type. (Double checked)
During the build process we don't use add. 

The build step mimics the stack-blitz example. So I guess what I will do is slowly add each custom fields, validators, hidden fields, wrappers, polymorph field, and anything else to the stackblitlz -- to test each thing tell I see the slowdown. (could be anyone of those)

**Comment by thorgod** [comment]
@aitboudad 

I do call two functions on build
```
loadData. .subscribe(x => {
this.ngForm.resetForm();  //Reset form hack for switching forms.
this.fields = x.formGroups; //Set fields
 this.model = x.model;  //Sets model.
this.form.markAsPristine(); 
});
```

**Comment by aitboudad** [comment]
normally `ngOnChanges` should be called only once which not the case in your app, I think you should try find the root cause of those calls.

**Comment by aitboudad** [comment]
it all related to the way we construct the `FormGroup` which update updateValidity for each field, I've made several PRs (#1499, #1500, #1501, #1504) which solve some edge cases, left 2 or 3 PRs to get done with the loading time.

In `V6` we'll remove `FormGroup` which I think will save an average of `~10%` for a repeatable field.

**Note**: all those changes will fix the loading time and not the typing performance, I think the only way to solve typing performance is: `onpush` changeDetection or virtual list (we can talk about it once done with the first part).

**Comment by thorgod** [comment]
@aitboudad  Thanks for the PR's I will take the master (looks like they are merged with master) and test it out this morning. 

 Its probably one thing causing it in my app, I hope to figure it out. Today I will add each part of the application to the stackblitz tell I figure out what is causing it.

**Comment by thorgod** [comment]
OK I completely stripped out my application, no custom validators, only one field type static html <h1>a</h1>.  Still takes around 20 seconds to load after data loaded. (This is with the latest master with your PRs) It looks like the '_executeValidators' might be doing something.

Next step I will look at the JSON structure, and start removing stuff from the JSON being sent maybe its something with Hidden/Required/Min/Max ect. 



![image](https://user-images.githubusercontent.com/13910170/55344411-15c0dd00-5462-11e9-92c3-e4a2e5be1a3e.png)

**Comment by aitboudad** [comment]
could you try to apply the following changes https://github.com/aitboudad/ngx-formly/commit/6164dd42403d990f2ece57712539fb6442880423 and let me know if it fix the loading or not

**Comment by thorgod** [comment]
Sure, I will do it right now

**Comment by thorgod** [comment]
Wow yes its significantly faster 3sec (Probably even faster when compiled without debug)  before it was around 20 sec .  (Now time to add all revert all the code back in and test it then)

**Comment by thorgod** [comment]
@aitboudad  Cool looking good, with all the back, its around ~7 Seconds from 20 seconds! 

Here is the performance graph. 
![image](https://user-images.githubusercontent.com/13910170/55348431-0a72af00-546c-11e9-895c-d0cf714bc281.png)

**Comment by aitboudad** [comment]
7 Seconds still too much but better than before :smile:

**Comment by thorgod** [comment]
A note: I have a ngx formly custom wrappers for every field in my application.

What can I do to help on my end?

**Comment by thorgod** [comment]
getting lunch be back in a hour.

**Comment by aitboudad** [comment]
using wrapper is not an issue, I don't think `formly` still cost much more time than is should be (the build part shouldn't exceed ~500ms in your case).

**Comment by thorgod** [comment]
Let me know if there is something I can do on my end. 
It will take sometime but I could make a stackblitz of my app portion that has ngx-formly, if you think that would be helpful?

**Comment by aitboudad** [comment]
can you provide me the runtime performance profiler of your project (that include the latest changes of formly) which you can save using the right click of the mouse
![Selection_002](https://user-images.githubusercontent.com/1753742/55354933-d70d4000-54be-11e9-812c-a6618329f75f.png)

**Comment by thorgod** [comment]
Actually it looks the same....so scrap my last comment. 

[Profile-20190401T132655.zip](https://github.com/ngx-formly/ngx-formly/files/3031217/Profile-20190401T132655.zip)

**Comment by aitboudad** [comment]
the built form takes ~1.5s and the rest (5s) somehow are consumed by destroying fields which probably related to something wrong with the usage of `formly`.
continue your debugging until you found out the root cause of that.

**Comment by thorgod** [comment]
OK sounds good I will keep removing stuff tell I find what it is.

I am using a wrapper to hide based on a condition could that be the cause?   (Well I figure it out, just guessing) 

```
<div [style.visibility]="hidden">
        <ng-template  #fieldComponent></ng-template>
      </div>
```

**Comment by aitboudad** [comment]
try to find where you've overridden the generated `field` (using the spread operator for example or a getter), because the generated part of the field will be loosed.

**Comment by thorgod** [comment]
I removed almost everything. Still getting 7 seconds, I will see if I can make a stackblitz example (Either today or tomorrow, because its so stripped down it should not be to hard to make) The build is fast =) Maybe its the debugger or something.

![image](https://user-images.githubusercontent.com/13910170/55361350-d7d8ae80-548b-11e9-9df1-44e2d34bc3fc.png)

**Comment by thorgod** [comment]
I updated the stackblitz with your changes and I think it matches closely to my application. (Tomorrow I will be adding my example JSON data / JSON Schema in that I was testing against) 

https://angular-ngx-formly-large-load-v4.stackblitz.io/

**Comment by thorgod** [comment]
@aitboudad   
https://stackblitz.com/edit/angular-ngx-formly-large-load-v5

OK I was able to repredouce time it takes to load on stackblitz. (I am going to double check I have all your latest changes but 98% sure) Nothing is complex just very simple input box. (I have multiple types however, I changed them all reference the same type 'OperatorComponent' a simple input box.)

Few notes:
Also if you are looking for a job I work for 'Thermo Fisher Scientific', I could put in a good word, or slight chance we can open a position. 

Would you like me to make an example for the documentation on how to use a more complex example of FieldArrayType, because I am sure there are a lot of people out there who want to use ngx-formly with tables? I would be also be happy to try and help move the FieldArray to its own module, it would be a good experience for me to learn. (Or at-least attempt)

**Comment by aitboudad** [comment]
now I have a clear idea, here is some additional changes that could improve the render time:
* use virtual list here is an example which use 1000 fields: https://stackblitz.com/angular/pxggyvvxeand
* apply this PR https://github.com/ngx-formly/ngx-formly/pull/1445/files#diff-d8598c62af5b27037ce9e85e1cd4591a, which enable lazy render
* avoid using wrapper for `operator` which apparently you don't use.
```patch
name: 'operator',
component: OperatorComponent,
-- wrappers: ['label']
++ wrappers: []
```

> Few notes:

leave the notes, until we get done with this issue :smiley:

**Comment by thorgod** [comment]
Oh I use wrappers! That was just to simplify the example.  But I will make those changes.

**Comment by thorgod** [comment]
Thanks for the help; however, overall around the same with the changes from the PR+Virtual List+Removing Wrapper. (Test with Master of ngx-formly) None the less I will test it on my app now. 

https://stackblitz.com/edit/angular-ngx-formly-large-load-v5    (Sorry forgot to fork before changing)

**Comment by aitboudad** [comment]
you've missed the commit https://github.com/aitboudad/ngx-formly/commit/6164dd42403d990f2ece57712539fb6442880423 which is not yet part of master branch

here is the result: https://stackblitz.com/edit/angular-ngx-formly-large-load-v5-63d3yz

for repeatable field I've added an `*ngIf`:
```ts
    <div *cdkVirtualFor="let field of field.fieldGroup; let i = index;">
      <formly-group *ngIf="!field.hide"
        [field]="field" style="display:flex;">
      </formly-group>
    </div>
```

**Comment by thorgod** [comment]
Ah thank you my bad on missing the  aitboudad@6164dd4 I did not know you wanted me to merge them all together.  I would of been happy to merge them just didn't get the memo. 

Looking good I will test it out my application as well

**Comment by thorgod** [comment]
Just a note with the experimental commit+master, it works but if I scroll to fast or navigate to a different page I get this error on my app. 

```
ERROR RangeError: Maximum call stack size exceeded
    at Array.forEach (<anonymous>)
    at FormlyField.refsUnsubscribe (ngx-formly-core.js:1203)
    at FormlyField.push.../../node_modules/@ngx-formly/core/fesm5/ngx-formly-core.js.FormlyField.ngOnDestroy (ngx-formly-core.js:1169)
    at callProviderLifecycles (core.js:22421)
    at callElementProvidersLifecycles (core.js:22389)
    at callLifecycleHooksChildrenFirst (core.js:22379)
    at destroyView (core.js:23441)
    at callViewAction (core.js:23567)
    at execEmbeddedViewsAction (core.js:23510)
    at destroyView (core.js:23439)
```

**Comment by aitboudad** [comment]
how much does it take now?

**Comment by thorgod** [comment]
Sorry didn't get back on the speed with the virtual list+pr+aitboudad/ngx-formly@6164dd4.  

In the application its significantly faster it takes around 7 seconds, without the changes 15-17 seconds. (Let me test without virtual list) 

[Profile-20190403T083139.zip](https://github.com/ngx-formly/ngx-formly/files/3039454/Profile-20190403T083139.zip)

**Comment by thorgod** [comment]
Removing virtual list, its around the same load time. 

Thanks for the help! I will look at your merge fix(core): improve performance for FieldArrayType #1511  to see what else you did.

**Comment by aitboudad** [comment]
> Sorry didn't get back on the speed with the virtual list

maybe you've missed :thinking:, IMO it should improve the speed.

All perf improvements are now published in `v5.0.1`, from 20s to 7s (using dev mode, so it should be less in prod env) is a huge step but I do believe that it can be improved much more than that.

The remaining part to improve is the render time, which depends on the number of components that are present in the dom, the following ways can fix that:

- use the virtual list (1000 fields, https://stackblitz.com/angular/pxggyvvxeand)
- use inline edit: https://stackblitz.com/edit/angular-bkawud
- use a simple `datatable` with crud actions to opens the form on demand using dialog or a popover

further steps:
- apply `ngIf="field.hide"` on all `formly-field` and `formly-group` to use lazy render.

**Comment by aitboudad** [comment]
> Few notes:
Also if you are looking for a job I work for 'Thermo Fisher Scientific', I could put in a good word, or slight chance we can open a position.

Right now I'm not looking for a job but I would be happy to apply when time will allow, thanks anyway :gift: 

> Would you like me to make an example for the documentation on how to use a more complex example of FieldArrayType, because I am sure there are a lot of people out there who want to use ngx-formly with tables? I would be also be happy to try and help move the FieldArray to its own module, it would be a good experience for me to learn. (Or at-least attempt)

PR welcomed!

**Comment by thorgod** [comment]
Thanks for the help! Going to test out a few of your suggestions but calling it quits after that. I am happy with what you were able to accomplish thanks again for the help.  


Just to give you some overall context:
Originally it took 66 seconds to load a biggest example form.
- 40 Seconds Server Side
   - AWS C# Serverless Lambda
   - RDS Microsoft SQL Database 
- 20 Seconds Angular.

I got it down to 
- 1-2 Seconds Server Side
   - AWS C# Serverless Lambda
   - RDS Microsoft SQL Database 
- 5-7 Seconds Angular. (Thanks to you)

![source (1)](https://user-images.githubusercontent.com/13910170/55515516-a1339d00-561f-11e9-913e-bdca65d5f98d.gif)

**Comment by aitboudad** [comment]
Hi @thorgod, I just want to let you know that I've pushed a performance improvement in [v5.5.12](https://github.com/ngx-formly/ngx-formly/releases/tag/v5.5.12) release that may reduce the 20 Seconds Delay.
if you're still working on that project please give it a try.

**Comment by Hyperonic** [comment]
using cdkVirtualFor calls onInit and onDestroy hooks repeatedly on scrolling. Is there any way to show large list but call onInit when the field appears once only?

**Comment by aitboudad** [comment]
@Hyperonic the better way is to use a simple datatable with a edit button to open the form on demand using dialog or a popover, rendering the entire form has a cost!

**Comment by ElesGergo** [comment]
Hello, I am facing performance issues myself. Big forms, hooks I have all the good stuff,even typeing to the form field is really slow , any updates on this one?  :/

**Comment by jainvandit25** [comment]
Hi @aitboudad  

I am creating a table using fieldType array around 6000 fields are there and my page becoming unresponsive when we push the data  to [model]=[model] . Could you please help . What can be done .to improve the perfomance

**Comment by aitboudad** [comment]
> Could you please help . What can be done .to improve the perfomance

see https://github.com/ngx-formly/ngx-formly/issues/2762#issuecomment-816828153

**Comment by jainvandit25** [comment]
> > Could you please help . What can be done .to improve the perfomance
> 
> see [#2762 (comment)](https://github.com/ngx-formly/ngx-formly/issues/2762#issuecomment-816828153)

I tried using *cdkVirtualFor but it doesnot provide any effect on my application . The page still show unresponsive

**Comment by aitboudad** [comment]
@jainvandit25 please create a separate ticket with a an example that reproduce the issue, to see if we can provide some hints in order to improve the performance for your app.
